File: CATEGORY/스택/BOJ/9012_괄호/9012.py
import sys

sys.stdin = open('input.txt','r')

def check_bracket(brackets):
    stack = []

    for bracket in brackets:
        if bracket == '(':
            stack.append(bracket)
        else:
            if len(stack) > 0:
                stack.pop()
            else:
                return 'NO'

    if len(stack) > 0:
        return 'NO'

    return 'YES'


T = int(input())

for tc in range(1,T+1):

    brackets = list(input())

    print(check_bracket(brackets))

File: javascript/BOJ/9012_괄호/9012.js
const fs = require('fs');
const fileStream = fs.createReadStream('input.txt');
const readline = require('readline');

const rl = readline.createInterface({
  input: fileStream,
  output: process.stdout,
})
let i = 0
let N = null;
let brackets = [];

function isVPS(bracket) {
  const stack = [];
  for (let i = 0; i < bracket.length; i++) {
    const b = bracket[i]
    if (b === '(') {
      stack.push("(")
    } else {
      if (stack.length > 0 && stack[stack.length - 1] === '(') {
        stack.pop()
      } else {
        return "NO"
      }
    }
  }

  return stack.length === 0 ? 'YES' : 'NO'
}

rl.on("line", (line) => {
  if (i === 0) {
    N = Number(line);
    i++
  } else brackets.push(line)
})
  .on("close", () => {
    brackets.forEach(bracket => console.log(isVPS(bracket)))
    process.exit()
  })